File: examples/js/end_to_end.mjs
// End-to-end xa11y example: drive the AccessKit test app from launch to teardown.
//
// This script is a complete, copy-pasteable starting point for writing your
// first xa11y test in JavaScript. It targets the AccessKit test app shipped
// with this repo (`test-apps/accesskit`) so it runs identically on Linux,
// macOS, and Windows.
//
// What it demonstrates:
//   * Launching a test app and polling the accessibility API until the OS
//     registers it (`App.byPid` with a `timeout`).
//   * Dumping the tree (`App.dump`) to discover the role and name of every
//     element before writing selectors.
//   * The `Locator` pattern with auto-waiting actions (`press`, `setValue`).
//   * Wait helpers: `waitVisible` (seconds) and `waitUntil` (milliseconds).
//   * Reading element fields (`role`, `name`, `actions`, `checked`).
//   * Subscribing to events with `app.subscribe()` and the
//     `Subscription.waitFor` helper.
//   * Tearing the subprocess down cleanly.
//
// Run from the repo root, after building the test app and JS bindings:
//
//   cargo build -p xa11y-test-app
//   (cd xa11y-js && npm install && npm run build:debug)
//   node examples/js/end_to_end.mjs

import { spawn } from 'node:child_process';
import { existsSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { setTimeout as sleep } from 'node:timers/promises';
import { fileURLToPath, pathToFileURL } from 'node:url';
import assert from 'node:assert/strict';

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = resolve(__dirname, '..', '..');

const isWindows = process.platform === 'win32';
const BINARY = resolve(
  REPO_ROOT,
  'target',
  'debug',
  isWindows ? 'xa11y-test-app.exe' : 'xa11y-test-app',
);

// Import the locally-built JS bindings without forcing a global install.
const xa11yUrl = pathToFileURL(resolve(REPO_ROOT, 'xa11y-js', 'index.js')).href;
const xa11y = await import(xa11yUrl);
const { App, SelectorNotMatchedError, PlatformError, TimeoutError } = xa11y;

const STARTUP_TIMEOUT_MS = 30_000;

async function waitForRegistration(pid) {
  const deadline = Date.now() + STARTUP_TIMEOUT_MS;
  let lastErr;
  while (Date.now() < deadline) {
    try {
      return await App.byPid(pid, { timeout: 1000 });
    } catch (err) {
      if (err instanceof SelectorNotMatchedError || err instanceof PlatformError) {
        lastErr = err;
        await sleep(200);
        continue;
      }
      throw err;
    }
  }
  throw new Error(`Test app (pid=${pid}) did not register within ${STARTUP_TIMEOUT_MS}ms: ${lastErr}`);
}

async function main() {
  if (!existsSync(BINARY)) {
    console.error(`Build the test app first: cargo build -p xa11y-test-app (looked at ${BINARY})`);
    process.exit(1);
  }

  // 1. Launch the test app. The example owns its subprocess lifecycle so a CI
  //    run never leaks processes between attempts.
  const proc = spawn(BINARY, [], { stdio: 'ignore' });
  try {
    // 2. Poll the accessibility API until the OS registers the new process.
    const app = await waitForRegistration(proc.pid);
    console.log(`App registered: ${app.name} (pid=${app.pid})`);

    // 3. Dump the tree once to discover the role/name of every element. Copy
    //    selectors out of this output instead of guessing.
    console.log('\n--- Tree (depth 4) ---');
    console.log(await app.dump(4));

    // 4. Locators auto-wait and re-resolve on every call, so they stay correct
    //    even if the UI mutates between operations.
    const submit = app.locator('button[name="Submit"]');
    await submit.waitVisible(5); // timeout in seconds for native wait helpers

    // 5. Read element fields via .element().
    const button = await submit.element();
    assert.equal(button.role, 'button');
    assert.equal(button.enabled, true);
    assert.ok(button.actions.includes('press'));

    // 6. Press the primary button.
    await submit.press();

    // 7. Drive a text input. `waitUntil` polls until the predicate is true —
    //    preferable to a fixed sleep. Timeout is milliseconds, mirroring
    //    other Node APIs.
    const nameField = app.locator('text_field[name="Name"]');
    await nameField.setValue('Ada Lovelace');
    try {
      await nameField.waitUntil((el) => el !== undefined && el.value === 'Ada Lovelace', {
        timeout: 2000,
      });
    } catch (err) {
      if (!(err instanceof TimeoutError)) throw err;
      // Some Linux AT-SPI adapters don't echo set_value back through the
      // tree; the call still went through. Print so the example is
      // transparent rather than silently passing.
      console.log('note: text value not echoed back via accessibility (adapter quirk)');
    }

    // 8. Toggle the checkbox via the `press` semantic verb and confirm the
    //    new state with `waitUntil`.
    const checkbox = app.locator('check_box[name="I agree to terms"]');
    const before = (await checkbox.element()).checked;
    await checkbox.press();
    await checkbox.waitUntil((el) => el !== undefined && el.checked !== before, {
      timeout: 2000,
    });
    const after = (await checkbox.element()).checked;
    console.log(`checkbox toggled: ${JSON.stringify(before)} -> ${JSON.stringify(after)}`);

    // 9. Iterate matching elements with `.elements()`.
    const buttons = await app.locator('button').elements();
    console.log(`discovered ${buttons.length} buttons total`);
    assert.ok(buttons.length >= 2);

    // 10. Subscribe to events and wait for one matching condition. The
    //     Subscription is an EventEmitter, and `waitFor` is the convenience
    //     wrapper for one-shot waits.
    const sub = await app.subscribe();
    try {
      const evPromise = sub.waitFor(
        (e) => e.target !== undefined && e.target.name === 'Submit',
        { timeout: 3000 },
      );
      await submit.press();
      const event = await evPromise;
      console.log(`observed event: ${event.eventType} on ${JSON.stringify(event.target.name)}`);
    } finally {
      sub.close();
    }

    console.log('\nOK — example completed successfully.');
  } finally {
    proc.kill('SIGTERM');
    await new Promise((r) => proc.once('exit', r));
  }
}

await main();

File: examples/python/end_to_end.py
"""End-to-end xa11y example: drive the AccessKit test app from launch to teardown.

This script is a complete, copy-pasteable starting point for writing your first
xa11y test in Python. It targets the AccessKit test app shipped with this repo
(``test-apps/accesskit``) so it runs identically on Linux, macOS, and Windows.

What it demonstrates:

* Launching a test app and polling the accessibility API until the OS registers it.
* Dumping the tree (``App.dump``) to discover the role and name of every element
  before writing selectors.
* The ``Locator`` pattern with auto-waiting actions (``press``, ``set_value``).
* Wait helpers: ``wait_visible``, ``wait_until``.
* Reading element fields (``role``, ``name``, ``actions``, ``checked``).
* Subscribing to accessibility events with ``app.subscribe`` and the
  ``Subscription.wait_for`` helper.
* Tearing the subprocess down cleanly.

Run from the repo root:

    cargo build -p xa11y-test-app
    python examples/python/end_to_end.py

Prerequisites: ``xa11y`` installed (``pip install -e xa11y-python``). On macOS,
the Python interpreter needs accessibility permission. On Linux, an X server
and an AT-SPI registry must be running.
"""

from __future__ import annotations

import subprocess
import sys
from pathlib import Path

import xa11y

REPO_ROOT = Path(__file__).resolve().parents[2]
BINARY = REPO_ROOT / "target" / "debug" / ("xa11y-test-app.exe" if sys.platform == "win32" else "xa11y-test-app")
STARTUP_TIMEOUT = 30.0


def main() -> int:
    if not BINARY.exists():
        sys.exit(f"Build the test app first: cargo build -p xa11y-test-app (looked at {BINARY})")

    # 1. Launch the test app as a subprocess. The example owns its lifecycle so
    #    that a CI run never leaks processes between attempts.
    proc = subprocess.Popen([str(BINARY)])
    try:
        # 2. Wait for the accessibility API to expose the app. ``by_pid`` polls
        #    internally until the timeout elapses; ``App.by_name`` is the
        #    alternative for apps you didn't launch yourself.
        app = xa11y.App.by_pid(proc.pid, timeout=STARTUP_TIMEOUT)
        print(f"App registered: {app.name} (pid={app.pid})")

        # 3. Dump the tree once to discover the role/name of every element.
        #    Copy a selector out of this output instead of guessing.
        print("\n--- Tree (depth 4) ---")
        print(app.dump(max_depth=4))

        # 4. Locators auto-wait and re-resolve on every operation, so they
        #    stay correct even if the UI mutates between calls.
        submit = app.locator('button[name="Submit"]')
        submit.wait_visible(timeout=5.0)

        # 5. Read element properties. Locators with a single match expose the
        #    underlying ``Element`` via ``.element()``.
        button = submit.element()
        assert button.role == "button", button.role
        assert button.enabled, "Submit should be enabled at startup"
        assert "press" in button.actions, button.actions

        # 6. Press the primary button.
        submit.press()

        # 7. Drive a text input. ``wait_until`` polls until the predicate is
        #    true — preferable to a fixed ``time.sleep``.
        #
        #    Some platform providers don't implement editable-text writes for
        #    every widget (e.g. Linux AT-SPI's AccessKit bridge doesn't expose
        #    ``EditableText`` — surfaced as ``ActionNotSupportedError``).
        #    Real apps usually expose it via Qt/GTK; the test app here is
        #    pure AccessKit, so we tolerate the error explicitly rather than
        #    swallowing it silently.
        name_field = app.locator('text_field[name="Name"]')
        try:
            name_field.set_value("Ada Lovelace")
        except xa11y.ActionNotSupportedError:
            print("note: set_value not supported by this provider (e.g. Linux AT-SPI on AccessKit)")
        else:
            try:
                name_field.wait_until(
                    lambda el: el is not None and el.value == "Ada Lovelace",
                    timeout=2.0,
                )
            except xa11y.TimeoutError:
                # Some providers accept set_value but don't echo it back
                # through the tree. The call still went through.
                print("note: text value not echoed back via accessibility (adapter quirk)")

        # 8. Toggle the checkbox via the ``press`` semantic verb and confirm
        #    the new state with ``wait_until``.
        checkbox = app.locator('check_box[name="I agree to terms"]')
        before = checkbox.element().checked
        checkbox.press()
        checkbox.wait_until(lambda el: el is not None and el.checked != before, timeout=2.0)
        print(f"checkbox toggled: {before!r} -> {checkbox.element().checked!r}")

        # 9. Iterate matching elements. ``.elements()`` returns all matches.
        buttons = app.locator("button").elements()
        print(f"discovered {len(buttons)} buttons total")
        assert len(buttons) >= 2

        # 10. Subscribe to events, trigger a press, and wait for the next
        #     event. In real code you would filter the predicate by
        #     ``e.event_type`` and/or ``e.target`` fields. Here we just
        #     demonstrate the API — pressing Submit mutates ``status_text``
        #     on the test app so an event is guaranteed to fire shortly after.
        with app.subscribe() as sub:
            submit.press()
            event = sub.wait_for(lambda _e: True, timeout=5.0)
            target_name = event.target.name if event.target else None
            print(f"observed event: {event.event_type} on {target_name!r}")

        print("\nOK — example completed successfully.")
        return 0
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()
            proc.wait(timeout=5)


if __name__ == "__main__":
    sys.exit(main())
